src/psyclone/psyir/transformations/acc_update_trans.py

```diff
@@ -314,54 +287,93 @@ def _add_update_directives(self, node_list):
+                # variable would appear in loop_sync as part of the outputs of
+                # any textually preceding kernels, we may simply place an
+                # update device directive inside the if statement for those
+                # variables in loop_sync. loop_sync also includes the outputs
+                # of this host region which are inputs of any textually
+                # preceding kernels but, even though these might not need to
+                # be updated here, i.e. conditionally, they can safely be.
                 if isinstance(sched.parent, Loop) or \
-                   isinstance(sched.parent, IfBlock) and idx == OUT:
-                    self._place_update(sched, update_pos, loop_sync, direction)
+                   isinstance(sched.parent, IfBlock) and acss_type == OUT:
+                    self._place_update(sched, update_pos, loop_sync, acss_type)
                     host_sig.difference_update(loop_sync)
 
                 # This schedule is the body of a routine and, at least until we
                 # can do interprocedural analysis, this is the end of the road.
                 if isinstance(sched, InvokeSchedule):
-                    self._place_update(sched, update_pos, host_sig, direction)
+                    self._place_update(sched, update_pos, host_sig, acss_type)
                     break
 
                 # Move up the code tree to the next deepest ancestor schedule.
-                while True:
+                child, sched = child.parent, sched.parent
+                while not isinstance(sched, Schedule):
                     child, sched = child.parent, sched.parent
-                    if isinstance(sched, Schedule):
-                        break
 
-    def _place_update(self, sched, update_pos, host_sig, direction):
+    def _sync_sig(self, dep_stmts, host_sig, acss_type):
+        '''
+        Retrieves, amongst the host region signatures in host_sig, those that
+        requires synchronisation because of accesses in the compute regions
+        within the statements in dep_stmts.
+
+        :param dep_stmts: list of statements which may include compute regions.
+        :type dep_stmts: List[:py:class:`psyclone.psyir.nodes.Statement`]
+        :param host_sig: access signature(s) that need to be synchronised \
+                         with the accelerator device.
+        :type host_sig: Set[:py:class:`psyclone.core.Signature`]
+        :param int acss_type: the access type from the point of view of the
+                              host, either IN or OUT.
+
+        '''
+        # If there is a statement (e.g. a call) among the dependent statements
+        # that may launch device kernels, we conservatively assume a dependency
+        # for all variables in the host region regardless of access type.
+        if any(stmt.walk(self._brk_nodes) for stmt in dep_stmts):
+            return host_sig.copy()
+
+        kern_sig = set()
```

Comment:
Please add a comment saying what this will collect, e.g. "Signatures that must be synchronised for compute kernels" or something.

Reply:
Done.